GPUQREngine/Source/GPUQREngine_Internal.cpp

```diff
@@ -20,32 +20,40 @@
 #include "GPUQREngine_Scheduler.hpp"
 #include "GPUQREngine_Stats.hpp"
 
+template class BucketList<int32_t>;
+template class BucketList<int64_t>;
+template class LLBundle<int32_t>;
+template class LLBundle<int64_t>;
 
+template class Scheduler<int32_t>;
+template class Scheduler<int64_t>;
```

Comment:
Instead of explicitly instantiating the templates here, it might be better to "forward declare" them with `extern template` instead.
That reduces compilation time. But more importantly, it might also avoid accidentally skipping a specialization from another compilation unit in this compilation unit.

Reply:
And instead of having to remember to "forward declare" those template classes manually in each compilation unit, it might be easier to move those `extern template` declaration to the respective headers.

Reply:
If they are not instantiated there where do you intend to instantiate them? Implicitly? I was getting linker errors using implicit.

Reply:
Ah. Sorry. I realized only now that the member functions of these template classes are defined across several different compilation units. I only skimmed through the code. But afaict, there are no template specializations. (Please, check if that's true.)

If the entirety of a template class was defined in one single compilation unit, it would have made sense to explicitly instantiate these classes in those respective compilation units. But this project seems to be structured differently.

I'd guess that explicitly instantiating these template classes in this compilation unit should be ok in this case. 👍

Sorry for my confusion.